pep-0484.txt

```diff
@@ -1015,6 +1015,40 @@ In such cases the default value may be specified as a literal
 ellipsis, i.e. the above example is literally what you would write.
 
 
+Annotating generator functions and coroutines
+---------------------------------------------
+
+The return type of generator functions can be annotated by
+the generic type ``Generator[yield_type, send_type,
+return_type]`` provided by ``typing.py`` module::
+
+  def echo_round() -> Generator[int, float, str]:
+      res = yield
+      while res:
+          res = yield round(res)
+      return 'OK'
+
+Coroutines introduced in PEP 492 are annotated with the same syntax as
+ordinary functions. However, the return type annotation corresponds to the
+type of ``await`` expression, not to the coroutine type::
+
+  async def spam(ignored: int) -> str:
+      return 'spam'
+
+  async def foo():
```

Comment:
I just updated the PEP. I think it's better to add, because the only known
example of a type checker does indeed ignore the body of unannotated
functions, so it's better to show examples that are fully annotated (except
when talking about annotated vs. unannotated).

On Thu, May 26, 2016 at 12:37 AM, Ivan Levkivskyi <notifications@github.com>
wrote:

> In pep-0484.txt
> <https://github.com/python/typing/pull/225#discussion_r64702492>:
>
> > +return_type]`` provided by ``typing.py`` module::
> > +
> > +  def echo_round() -> Generator[int, float, str]:
> > +      res = yield
> > +      while res:
> > +          res = yield round(res)
> > +      return 'OK'
> > +
> > +Coroutines introduced in PEP 492 are annotated with the same syntax as
> > +ordinary functions. However, the return type annotation corresponds to the
> > +type of ``await`` expression, not to the coroutine type::
> > +
> > +  async def spam(ignored: int) -> str:
> > +      return 'spam'
> > +
> > +  async def foo():
>
> In the initial version I had this -> None, but then removed to simplify
> the example. If I understand correctly this sentence in the PEP "Any
> function without annotations should be treated as having the most general
> type possible, or ignored, by any type checker", then it is up to
> type-checker whether to type check the body of foo here. I do not have
> any preference here, if you think it is better to add -> None, then
> should I make another PR (this one is already merged), of you will add it?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/python/typing/pull/225/files/30ec2751a96be037a02879b6c551b7a00d05a596#r64702492>
>



-- 
--Guido van Rossum (python.org/~guido)